### ISSUE TITLE
Inconsistent expansion of `&`

### ISSUE THREAD

**Issue by arimah** [styled-components/styled-components#3265]
TL;DR: `&` sometimes expands to the static component class name, sometimes to the dynamic props-based class name.

### System:
 - OS: Windows 10 10.0.19041
 - CPU: (8) x64 Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
 - Memory: 4.97 GB / 15.88 GB
### Binaries:
 - Node: 12.16.1 - C:\Program Files\nodejs\node.EXE
 - npm: 6.13.4 - C:\Program Files\nodejs\npm.CMD

... plus styled-components@5.2.0 and babel-plugin-styled-components@1.11.1, which didn't show up in the output for some reason.

## Reproduction

https://codesandbox.io/s/styled-components-inconsistent-expansion-5qeqt?file=/index.js

Use the radio buttons to toggle different class names on `Outer`: both `first` and `second` should cause only the second `Inner` to become highlighted in green, but `first` ends up highlights both.

## Steps to reproduce

The following component is problematic:

```js
const Outer = styled.div`...`;
const Inner = styled.div`
  ${Outer}.first &,
  ${Outer}.second & {
    background-color: ${(p) => (p.highlight ? "green" : "gray")};
  }
`;
```

The two `&`s in `Inner`'s selector expand to different class names

## Expected Behavior

The `&`s in the selector should both expand either to the static component class name, or to the dynamic props-based class name. In other words, either of these selector shapes would be acceptable:

```
.sc-Outer.first .sc-Inner, .sc-Outer.second .sc-Inner
```

or

```
.sc-Outer.first .dynamicInner, .sc-Outer.second .dynamicInner
```

Although the former seems distinctly less useful: being unable to use dynamic styles in this way would largely negate the point of styled-components.

## Actual Behavior

The first `&` expands to the static class name; the second to the dynamic class name. My browser generates the following styles:

```css
.sc-bkbjAj.first .sc-iemXMA, .sc-bkbjAj.second .eqeJWx {
    background-color: green;
}
.sc-bkbjAj.first .sc-iemXMA, .sc-bkbjAj.second .fZFUcP {
    background-color: gray;
}
```

Related to #3236?

---

Slightly off-topic, but: was 5.2.0 an accidental release? The new features match what's under "unreleased" in the changelog.

**Comment by woshizja** [comment]
I have the same problem，the code：
``` javascript
const TestComponent = styled.div`
  &+& {
    padding-left: ${p => (p.isSomething ? '9px' : undefined)};
  };
`;
```
use TestComponent:
``` javascript
const [isSomething, setIsSomething] = useState(true);
  const handleClick= () => {
    setIsSomething(!isSomething);
};

const Component = (
<>
<TestComponent isSomething={isSomething}>
    TestComponent 111
</TestComponent>
<TestComponent isSomething={isSomething}>
    TestComponent 222
</TestComponent>
<button onClick={handleClick}>do something</button>
</>
)
```
the css will always be whether isSomething is true or false :
``` css
.sc-fWWZrV+.sc-fWWZrV {
    padding-left: 9px;
}
```

**Comment by eramdam** [comment]
Had the same issue when upgrading, you might want to look at the comments in this thread https://github.com/styled-components/styled-components/issues/3244#issuecomment-687676703

TL;DR: You can replace `&` by `&&` as a workaround

**Comment by gaauwe** [comment]
I've been keeping an eye on this issue for some time now and since there is no clear update I wanted to give my 2 cents. After updating from v5.1.1 to v5.2.0 we have the same problem, so for now we've put updating to the newest version on hold. Without changing anything besides the styled-components version, our styling behaves differently in some places. For example we have multiple checkbox elements, that have custom styling if they're inside a certain div and have a different backgrounds based on their checked state:

```
${Section} & {
    transition: background-color 250ms ease-in-out;
    background: ${p => p.state.checked ? p.theme.colors.red : p.theme.colors.blue};
}
```

After updating to v5.2.0 this is causing some weird effects. If you select one specific checkbox, the background of all the checkboxes in that div change. Also de-selecting a checkbox doesn't change the styling back. So when the page is first loaded, all the checkboxes have a blue background. If you then select one checkbox, all the backgrounds will change to red. When you de-select a checkbox the background stays red. We do see the styling being overwritten in Chrome DevTools, instead of adding/using a separate class for the changed element:

```
.sc-1bijn1h-1 {
    transition: background-color 200ms ease-in-out;
    background: rgb(235, 214, 221);   
}

.sc-1bijn1h-1 {
    transition: background-color 200ms ease-in-out;
    background: rgb(200, 223, 241);
}
```

We can use the workaround of replacing `&` with `&&` (as described here [#3244 (comment)](https://github.com/styled-components/styled-components/issues/3244#issuecomment-687676703)) but as far as I understand this is a workaround and not a solution, correct me if I'm wrong though. There are some other issues that seem related to this problem: [#3273](https://github.com/styled-components/styled-components/issues/3273) and [#3288](https://github.com/styled-components/styled-components/issues/3288). Where [#3288](https://github.com/styled-components/styled-components/issues/3288) is the most simple case, because only one property changes and there seems to be no usages of `&`.

It seems related to static classes vs dynamic classes and as mentioned somewhere else it has maybe something to do with this change: [#3236](https://github.com/styled-components/styled-components/pull/3236)? If more reproducible code would be helpful, I would be glad to help in any way I can. Otherwise if in the end using `&&` is the intended way forward, I think it would be best to mark this as a breaking release.

**Comment by garyee** [comment]
5.2.0 also broke my UI.

I have a component called box, which can be grouped by using the boxGroup component.
A group can be vertical or horizontal (flex-direction). Here is the catch: they can be nested.
I used the sibling selector to 'configure' the margins.

Since styled-components did change the sibling-selector behaviour the margins are broke:

I created a [codesandbox](https://stackoverflow.com/questions/9185434/using-svg-as-background-image), with box and box group (sorry for the colors!)
- A box is red
- A boxGroup is blue(horizontal) or yellow(vertical)

There are two problems:
1. The red boxes should be surrounded by only one color (box + boxGroup get the wrong margin)
2. The red boxes in the nested group get two margins (left & top)

**Comment by probablyup** [comment]
this should help clear it up https://github.com/styled-components/styled-components-website/pull/743

**Comment by gaauwe** [comment]
We updated our code based on the new documentation linked above, and it indeed fixed the issue. It's much clearer in my opinion on how `&` should be used, thanks for that!

We did however notice a decrease in performance since we updated to v5.2.1. We keep track of our page speed with lighthouse tests that run every hour. Since we released our code with styled-components v5.2.1 we noticed a significant increase in TBT:

![image](https://user-images.githubusercontent.com/1271226/102331202-d0707600-3f8a-11eb-9d8a-5b27d1524cd4.png)

One particular thing we noticed is that it seems like CSS is getting duplicated. For example, we have a few different checkboxes with the same base style, but with extra styling once they're checked. This results in the following:

```
// Unique style of a checked checkbox.
.Page__Page-uxmfyn-1 .erCMAz.erCMAz {
    background-color: red;
}

// Base style of a checkbox (checked element).
.erCMAz {
    -webkit-appearance: none;
    -moz-appearance: none;
    appearance: none;
    box-sizing: border-box;
    display: inline-block;
    width: 16px;
    height: 16px;
    cursor: pointer;
    border: 1px solid #cccccc;
}

// Base style of the other checkboxes.
.kzyxeg {
    -webkit-appearance: none;
    -moz-appearance: none;
    appearance: none;
    box-sizing: border-box;
    display: inline-block;
    width: 16px;
    height: 16px;
    cursor: pointer;
    border: 1px solid #cccccc;
}
```

This corresponds to the following DOM:

```
<div class="Page__Page-uxmfyn-1">
    <input type="checkbox" name="One" class="InputCheckboxstyle__StyledInputCheckbox-a59m8o-2 erCMAz" value="One" checked>
</div>
```

@probablyup Are you aware of these performance effects and could it have something to do with the duplicate CSS?

**Comment by piter2k1** [comment]
Hi, I have the same problem as above.

I made two examples, one with version 5.1.1 (which is working good) and with bad 5.3.0 version (all > 5.1.1 not working good).

Bad example (5.3.0):
https://codesandbox.io/s/styled-components-530-3ltkl

Working example (5.1.1):
https://codesandbox.io/s/styled-components-511-r7uct

```jsx
const StyledBox = styled.div`
  width: 100px;
  height: 20px;
  background: lightgray;
  margin: 10px;

  .foo & {
    width: ${({ width }) => (width ? `${width}px` : "100px")};
  }
`;
```

**Comment by bensaufley** [comment]
Not sure whether to post on the PR or here, but here has more conversation so I'll post here:

`&&` as "reference to an instance" does not behave the same as `&` does as "reference to all instances"—as the doc update notes, `&&` produces `.hash.hash`. That means that if one _needs_ to use `&&` for instance-based style changes, but then has to override them somewhere (e.g. `${Foo} ${Bar}`), you begin to find yourself in a CSS specificity war.

That update to the documentation is not _wrong_ in describing the behavior but it does feel a bit like it's describing a bug as a feature? As others have mentioned in this thread, the first `&` in a comma-separated list of selectors will reference "all instances" but the second does not. That doesn't seem to be expected behavior—it's not even _internally_ consistent.

For me, just using one `&` inside a helper that produces `${Foo} &` means I always need to use `&&` _in case_ the helper is referenced inside a dynamic component, which puts me into the aforementioned specificity war pretty quickly: this one selector is now three orders of specificity deep, and I'll need _four_ to override it. I know there will be an argument that SC is built not to have a lot of cross-reference but even a single reference will begin this pattern.

**Comment by davidcollingwood** [comment]
Hi @kitten and @probablyup,

This issue has prevented our team from upgrading (we share @bensaufley's concerns above around CSS specificity regarding the use of `&&`). We've delayed as long as possible while waiting for a fix, but we're now in a position where styled-components@5.1.1 is no longer compatible with React 18 (console warnings about "Invalid use of hooks").

Can this issue please be addressed as part of v6?

Potential solutions to consider:
* `&` to always refer to the dynamic instance class when generating conditional styles
* `&` to always refer to the dynamic instance class
* A new special character (eg. `%`) to refer to the dynamic instance class explicitly
* Remove the "precedence boost" feature (ie. `&&` creates only a single dynamic class name instead of two)
* Any other ideas you can think of?

Would really appreciate your thoughts and opinions on whether any of the above ideas are achievable.
Thanks!

**Comment by itayganor** [comment]
A workaround I sometimes tend to do is:

```js
const Button = styled.button`
    // ... styles for common buttons

    &:is(${Card} *) {
        // ... styles for card buttons
    }
`;
```

`&:is(${Card} *)` is theoretically equivalent to `${Card} &` and [is greatly supported](https://caniuse.com/css-matches-pseudo).

**Comment by mmelnic** [comment]
still have this problem on 5.3.6

my solution is to use fake first selector and then normal ones (not to use && to not have double .class name)
`
  ${TownBuildingBaseContainer}.fake,
  ${TownBuildingBaseContainer}[data-hovered='true'] > &,
  ${TownBuildingBaseContainer}[data-character-selected='true']:hover > & {
    background-image: url(${props => props.backgroundUrlHovered});
  }

  ${TownBuildingBaseContainer}.fake,
  ${TownBuildingBaseContainer}[data-hovered='true'] &,
  ${TownBuildingBaseContainer}:hover & {
    opacity: 100%;
  }

  ${TownBuildingBaseContainer}.fake,
  ${TownBuildingBaseContainer}:active[data-inactive='false'] & {
    scale: 96%;
  }
`

wish the bug will be fixed at least on 6 version release
because I faced this bug every 2-3 months somewhere in a project and thought that it is my selector problem, but found that it is popular problem just now

**Comment by quantizor** [comment]
Closing — this issue is stale (2+ years without resolution) and relates to an older version or has been superseded by v6.x changes. If this is still reproducible on the latest v6.x release, please open a new issue with a minimal reproduction.
